### DISCUSSION TITLE
Non-blocking I2C driver

### DISCUSSION THREAD

**Discussion by mbait** [modm-io/modm#1058] [Ideas]
Would you accept a patch that enables non-blocking I/O in the I2C driver? I had implemented a working prototype for AVR before I came across this project, and now I'm thinking to start contributing to modm rather than continuing my sole efforts. I2C is really slow compared to what one can squeeze out of AT Mega (16MHz vs 400kHz). The proposed solution would add non-blocking read/write routines with a callback argument. One issue with this approach is that a straightforward solution implies calling the callback function straight out of the ISR, and that's not always good - unaware users may add heavy operations or nested calls to the I2C driver.

**Comment by salkinium** [comment]
I'm a little confused, because afaik modm already implements I2C non-blocking via the I2cTransaction mechanism and resumable functions or fibers? What am I missing?

**Reply by mbait** [reply]
I've only spent a day browsing throught the sources, so might overlook something. Let me give it a second try.
